controllers/propagator/replicatedpolicy_setup.go

```diff
@@ -1,48 +1,76 @@
 package propagator
 
 import (
+	"context"
+	"sync"
+	"time"
+
 	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/workqueue"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/event"
 	"sigs.k8s.io/controller-runtime/pkg/handler"
 	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 	"sigs.k8s.io/controller-runtime/pkg/source"
 
 	policiesv1 "open-cluster-management.io/governance-policy-propagator/api/v1"
 	"open-cluster-management.io/governance-policy-propagator/controllers/common"
 )
 
-func (r *ReplicatedPolicyReconciler) SetupWithManager(mgr ctrl.Manager, additionalSources ...source.Source) error {
-	builder := ctrl.NewControllerManagedBy(mgr).
+func (r *ReplicatedPolicyReconciler) SetupWithManager(
+	mgr ctrl.Manager, dynWatcherSrc source.Source, updateSrc source.Source,
+) error {
+	return ctrl.NewControllerManagedBy(mgr).
 		Named("replicated-policy").
 		For(
 			&policiesv1.Policy{},
-			builder.WithPredicates(replicatedPolicyPredicates()),
-		)
-
-	for _, source := range additionalSources {
-		builder.WatchesRawSource(source, &handler.EnqueueRequestForObject{})
-	}
-
-	return builder.Complete(r)
+			builder.WithPredicates(replicatedPolicyPredicates(r.ResourceVersions))).
+		WatchesRawSource(
+			dynWatcherSrc,
+			// The dependency-watcher could create an event before the same sort of watch in the
+			// controller-runtime triggers an update in the cache. This tries to ensure the cache is
+			// updated before the reconcile is triggered.
+			&delayGeneric{
+				EventHandler: &handler.EnqueueRequestForObject{},
+				delay:        time.Second * 3,
```

Comment:
This is fine for now, but I should be able to address this case with my refactoring of the template library.

Reply:
That would be great!